Improve Amphp performance

dbalabka opened the pull request:
Amphp should implement queuing approach similar to React. 
WIP. It didn't yet perform manual testing. So not sure that this works.

dbalabka committed:
Improve Amphp performance

Insolita commented:
Hi, thanks a lot!

Insolita commented:
Merged to the "[next](https://github.com/Insolita/php-async-benchmarks/commit/2a4906ecde375ec6dfae537a1ea1babca46e09ab)" branch

Insolita commented:
Huh, make it workable https://github.com/Insolita/php-async-benchmarks/blob/next/src/clients/Amphp.php

But it still slow.   ¯\_(ツ)_/¯
Probably I haven't some of the necessary libs?

kelunik commented:
The queue is built-in depending on the pool you choose. I'll send a PR. :-)

dbalabka commented:
@Insolita I have adjusted code and fix the bugs. 

@kelunik I tried to get how to use amp/parallel Pools but it is not clear from the documentation. In my case, I have implemented a simple queuing based on SplQueue and pool of workers (promises).

kelunik commented:
@dbalabka Parallel isn't needed, but `amphp/sync` can be helpful, see #2. It's quite a bit faster than Guzzle for me now.

dbalabka commented:
@kelunik I just reviewed the code. In my opinion, semaphores do not look so straight forward. Could you please compare my implementation (I created another PR due to this one closed):
https://github.com/Insolita/php-async-benchmarks/pull/3/files 
with yours and provide your feedback. It would be very valuable.

kelunik commented on src/clients/Amphp.php:
```diff
@@ -57,49 +61,48 @@ private function initUrls()
+                continue;
             }
+            // wait when some task will be accomplished
+            [$errors, $values] = yield Promise\some($pool);
```
Using `some` here will register a callback to each promise every iteration, which will negatively affect performance and memory usage.

dbalabka replied:
It is fixed. Please see new PR:
https://github.com/Insolita/php-async-benchmarks/pull/3/files

dbalabka replied:
I use `Promise\first()`, so a similar problem will be there

Insolita commented:
Oh, guys, thanks a lot!  
I'm confused now...  Both versions work good, but it so different

#2 
![kelunik](https://user-images.githubusercontent.com/1847402/80916286-53d9b000-8d8a-11ea-9f64-81619be573bc.png)
 #3 
![dbalabka](https://user-images.githubusercontent.com/1847402/80916287-550add00-8d8a-11ea-85e0-83df8ac8ed92.png)

kelunik commented:
@Insolita Seems like the other clients have similar timing differences in the README?

kelunik commented:
```
php bin/bench -t=amp --batch=200 --conk=50 --iters=2
0 - 31.3855509758
1 - 10.08562707901
```

I just got this locally, it's likely due to DNS being cached on the second run.

kelunik commented:
Most time is spend inside DNS / connection logic:

```
cat temp/amp/requests.har| jq '.log.entries[].timings.connect' | datamash mean 1
827,52331606218
```

You can get that `requests.har` by adding `->intercept(new LogHttpArchive($this->tempDir . '/requests.har'))` on `HttpClientBuilder`.
